garak/generators/multi_endpoint_rest.py

```diff
@@ -0,0 +1,458 @@
+import json
+import logging
+import requests
+requests.packages.urllib3.disable_warnings() 
+import backoff
+import jsonpath_ng
+import os
+from typing import List, Union
+from jsonpath_ng.exceptions import JsonPathParserError
+
+
+from garak import _config
+from garak.generators.base import Generator
+from garak.exception import APIKeyMissingError, RateLimitHit
+
+
+class MultiEndpointGenerator(Generator):
+    """A REST generator that sends a POST request and retrieves the response with a subsequent GET request.
+    Example configuration json;
+    {
+        "multi_endpoint_rest": {
+            "MultiEndpointGenerator": {
+                "name": "example service",
+                "post_uri": "https://example.ai/llm",
+                "post_headers": {
+                    "X-Authorization": "$KEY"
+                },
+                "post_req_template_json_object": {
+                    "text": "$INPUT"
+                },
+                "post_response_json": true,
+                "post_response_json_field": "job_id",
+                "get_uri": "https://example.ai/llm",
+                "get_headers": {
+                    "X-Authorization": "$KEY"
+                },
+                "get_req_template_json_object": {
+                    "text": "$INPUT"
+                },
+                "post_response_json": true,
+                "post_response_json_field": "text"
+            }
+        }
+    }
```

Comment:
This example config does not match the class expectations.

``` json
    {
        "multi_endpoint_rest": {
            "MultiEndpointGenerator": {
                "name": "example service",
                "first_stage_uri": "https://example.ai/llm",
                "first_stage_headers": {
                    "X-Authorization": "$KEY"
                },
                "first_stage_req_template_json_object": {
                    "text": "$INPUT"
                },
                "first_stage_response_json": true,
                "first_stage_response_json_field": "job_id",
                "second_stage_uri": "https://example.ai/llm",
                "second_stage_headers": {
                    "X-Authorization": "$KEY"
                },
                "second_stage_req_template_json_object": {
                    "text": "$INPUT"
                },
                "second_stage_response_json": true,
                "second_stage_response_json_field": "text"
            }
        }
    }
```

`first_stage_` & `second_stage_` prefix are very generic would there be value in a more specific network concepts focused prefix like `send_` & `recv_`?

``` json
    {
        "multi_endpoint_rest": {
            "MultiEndpointGenerator": {
                "name": "example service",
                "send_uri": "https://example.ai/llm",
                "send_headers": {
                    "X-Authorization": "$KEY"
                },
```

Another option might be to organize the config in line with reuse of the existing rest generator.

``` json
    {
        "multi_endpoint_rest": {
            "MultiEndpointGenerator": {
                "first_stage": {                    
                    "name": "request service",
                    "uri": "https://example.ai/llm",
                    "method": "post",
                    "headers": {
                        "X-Authorization": "$KEY",
                    },
                    "req_template_json_object": {
                        "text": "$INPUT"
                    },
                    "response_json": true,
                    "response_json_field": "text"
                },
                "second_stage": {
                    "name": "response service",
                    "uri": "https://example.ai/llm",
                    "method": "post",
                    "headers": {
                        "X-Authorization": "$KEY",
                    },
                    "req_template_json_object": {
                        "text": "$INPUT"
                    },
                    "response_json": true,
                    "response_json_field": "text"
                }
            }
        }
    }
```

Reply:
The first and second stage variable names were intentionally generic for flexibility, as it can perform any two calls to any two systems. A post and delete for example, or two puts (one to make a change and the other to revert it).